### PR TITLE
Company List Window

### PR DESCRIPTION
**Pull request by Svelbeard**

### CODE REVIEW

File: src/openloco/companymgr.cpp [hunk 1]
```diff
@@ -140,6 +140,21 @@ namespace openloco::companymgr
...
+        regs.esi = (int32_t)get(id);
+        call(0x00438047, regs);
+
+        owner_status ownerStatus;
+
+        ownerStatus.string = regs.bx;
+        ownerStatus.argument1 = regs.ecx;
+        ownerStatus.argument2 = regs.edx;
+
+        return ownerStatus;
```
Comment (AaronVanGeffen) [inline src/openloco/companymgr.cpp:155]: This variable is on the stack, and goes out of scope when this function returns.

Comment (Svelbeard) [reply src/openloco/companymgr.cpp:155]: Should I pass in a pointer to an `owner_status` variable and not return a value instead?

Comment (duncanspumpkin) [reply src/openloco/companymgr.cpp:155]: It's fine it's sending back a copy (probably is actually a move not a copy but that's besides the point) not a pointer to the stack.

File: src/openloco/windows/CompanyList.cpp [hunk 2]
```diff
@@ -1,18 +1,1732 @@
...
+
+            self->enabled_widgets = tabInfo.enabledWidgets;
+            self->holdable_widgets = 0;
+            self->event_handlers = tabInfo.events;
+            self->activated_widgets = 0;
+            self->widgets = tabInfo.widgets;
+
+            self->invalidate();
+
+            switch (self->current_tab + widx::tab_company_list)
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:1460]: can just use widgetIndex as you have already have this computed

File: src/openloco/windows/CompanyList.cpp [hunk 3]
```diff
@@ -1,18 +1,1732 @@
...
+                    return orderByPerformance(lhs, rhs);
+
+                case SortMode::Value:
+                    return orderByValue(lhs, rhs);
+            }
+
+            return false;
+        }
+
+        // 0x00499E0B
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:263]: `0x437AE2`

File: src/openloco/windows/CompanyList.cpp [hunk 4]
```diff
@@ -1,18 +1,1732 @@
...
+
+                companymgr::get(chosenCompany)->challenge_flags |= company_flags::sorted;
+
+                if (chosenCompany != self->row_info[self->row_count])
+                {
+                    self->row_info[self->row_count] = chosenCompany;
+                    shouldInvalidate = true;
+                }
+
+                self->row_count += 1;
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:303]: `++`

File: src/openloco/windows/CompanyList.cpp [hunk 5]
```diff
@@ -1,18 +1,1732 @@
...
+            gfx::draw_string_494B3F(*dpi, xPos, yPos, colour::black, string_ids::black_stringid, &args);
+        }
+
+        // 0x00435EA7
+        static void draw_scroll(window* self, gfx::drawpixelinfo_t* dpi, uint32_t scrollIndex)
+        {
+            //registers regs;
+            //regs.esi = (int32_t)self;
+            //regs.edi = (int32_t)dpi;
+            //call(0x00435EA7, regs);
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:471]: remove the commented code.

File: src/openloco/windows/CompanyList.cpp [hunk 6]
```diff
@@ -1,18 +1,1732 @@
...
+
+            auto colour = colour::get_shade(self->colours[1], 3);
+            gfx::clear(*dpi, colour * 0x01010101);
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:474]: `gfx::clear_single`

File: src/openloco/windows/CompanyList.cpp [hunk 7]
```diff
@@ -1,18 +1,1732 @@
...
+
+            auto y = 0;
+            for (auto i = 0; i < self->var_83C; i++)
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:477]: you can have multiple items in the last statement of the if using the comma operator.
`for (auto i = 0; i < var_83C; i++, y+=25)`. You could also just do `y = i * 25`

File: src/openloco/windows/CompanyList.cpp [hunk 8]
```diff
@@ -1,18 +1,1732 @@
...
+            {
+                auto yPos = y + 25;
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:479]: yPos represents the top of the row and y represents the bottom. Rename the variables to make this easier to follow.

File: src/openloco/windows/CompanyList.cpp [hunk 9]
```diff
@@ -1,18 +1,1732 @@
...
+                    args.push(ownerStatus.string);
+                    args.push(ownerStatus.argument1);
+                    args.push(ownerStatus.argument2);
+
+                    gfx::draw_string_494BBF(*dpi, 175, y + 7, 208, colour::black, stringId, &args);
+                }
+
+                auto performanceStringId = string_ids::performance_index;
+
+                if (company->challenge_flags & (company_flags::increased_performance | company_flags::decreased_performance))
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:538]: change this to a if decreased and if increased.

File: src/openloco/windows/CompanyList.cpp [hunk 10]
```diff
@@ -1,18 +1,1732 @@
...
+        }
+
+        // 0x00437AB6
+        static void sub_437AB6(window* self)
+        {
+            self->row_count = 0;
+
+            for (auto& company : companymgr::companies())
+            {
+                company.challenge_flags &= ~(1 << 3);
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:576]: missing the check on if company is clear

File: src/openloco/windows/CompanyList.cpp [hunk 11]
```diff
@@ -1,18 +1,1732 @@
...
+
+        if (window == nullptr)
+        {
+            gfx::ui_size_t windowSize = { 640, 272 };
+
+            window = WindowManager::createWindow(WindowType::companyList, windowSize, 0, &company_list::events);
+        }
+
+        //window->enabled_widgets = company_list::enabledWidgets;
+        //window->current_tab = 0;
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:634]: remove dead code

File: src/openloco/windows/CompanyList.cpp [hunk 12]
```diff
@@ -1,18 +1,1732 @@
...
+        window->frame_no = 0;
+        window->saved_view.clear();
+        window->flags |= window_flags::resizable;
+        window->var_83C = 0;
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:638]: missing setting sort mode to `2`

File: src/openloco/windows/CompanyList.cpp [hunk 13]
```diff
@@ -1,18 +1,1732 @@
...
+        window->row_hover = -1;
+
+        company_list::sub_437AB6(window);
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:641]: You've changed the behaviour of the window just slightly. Before it wouldn't reset as many things when clicking the window open again.

File: src/openloco/windows/CompanyList.cpp [hunk 14]
```diff
@@ -1,18 +1,1732 @@
...
+                args.push(company.name);
+
+                gfx::draw_string_494BBF(*dpi, x + 6, y, 94, colour::black, stringId, &args);
+
+                y += 10;
+                companyCount++;
+            }
+        }
+
+        static void drawGraphAndKey(window* self, gfx::drawpixelinfo_t* dpi)
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:1688]: add an offset `// 0x004365E4`

File: src/openloco/windows/CompanyList.cpp [hunk 15]
```diff
@@ -1,18 +1,1732 @@
...
+            _graphXAxisRange = totalMonths;
+            _dword_113DD7C = 1;
+            _byte_113DD99 = 1;
+
+            common::drawGraph(self, dpi);
+
+            if (self->var_854 != 0)
+            {
+                auto i = 0;
+                while (utility::bitscanforward(self->var_854) != _graphItemId[i])
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:1701]: perform the scan first and also check for `-1`

File: src/openloco/windows/CompanyList.cpp [hunk 16]
```diff
@@ -1,18 +1,1732 @@
...
+                if (company.empty())
+                    continue;
+
+                auto companyColour = companymgr::get_company_colour(company.id());
+                auto colour = colour::get_shade(companyColour, 6);
+                auto stringId = string_ids::small_black_string;
+
+                if (self->var_854 & (1 << companyCount))
+                {
+                    stringId = string_ids::small_white_string;
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:1670]: This appears to be different to vanilla.

File: src/openloco/windows/CompanyList.cpp [hunk 17]
```diff
@@ -1,18 +1,1732 @@
...
+            for (auto& company : companymgr::companies())
+            {
+                if (company.empty())
+                    continue;
+
+                auto companyColour = companymgr::get_company_colour(company.id());
+                auto colour = colour::get_shade(companyColour, 6);
+                auto stringId = string_ids::small_black_string;
+
+                if (self->var_854 & (1 << companyCount))
```
Comment (duncanspumpkin) [inline src/openloco/windows/CompanyList.cpp:1668]: 854 is the currently selected keys?

File: src/openloco/windows/CompanyList.cpp (hunk 13 above)
Comment (Svelbeard) [reply src/openloco/windows/CompanyList.cpp:641]: Sorry can you elaborate? I'm not sure I can see the difference.

File: src/openloco/windows/CompanyList.cpp (hunk 17 above)
Comment (Svelbeard) [reply src/openloco/windows/CompanyList.cpp:1668]: Yep, looks like it

File: src/openloco/windows/CompanyList.cpp (hunk 13 above)
Comment (duncanspumpkin) [reply src/openloco/windows/CompanyList.cpp]: In the previous version the sort mode and a few other variables would not be reset when you reopen the window.

Comment (Svelbeard) [reply src/openloco/windows/CompanyList.cpp]: Oh right, so the previous commit fixed that?

Comment (duncanspumpkin) [reply src/openloco/windows/CompanyList.cpp]: No you still need to move some of these setting of values up into the new window branch of the code

**Comment by duncanspumpkin** [comment]
Could you rebase this then its good to go.

**Pull request merged**